dds/idl/field_info.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ *
+ *
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include "field_info.h"
+#include <sstream>
+
+using namespace AstTypeClassification;
+
+FieldInfo::EleLen::EleLen(FieldInfo& af) : ele_(af.as_base_), len_(af.n_elems_)
+{
+}
+
+bool FieldInfo::EleLen::operator<(const EleLen& o) const
+{
+  return ele_ < o.ele_ || (ele_ == o.ele_ && len_ < o.len_);
+}
+
+const std::string FieldInfo::scope_op = "::";
+
+bool FieldInfo::cxx11()
+{
+  return be_global->language_mapping() == BE_GlobalData::LANGMAP_CXX11;
+}
+
+std::string FieldInfo::at_pfx() {
+  return cxx11() ? "T_" : "_";
+}
+
+std::string FieldInfo::get_type_name(AST_Type& field)
+{
+  std::string n = scoped(field.name());
+  if (!field.anonymous()) {
+    return n;
+  }
+  n = n.substr(0, n.rfind(scope_op) + 2) + at_pfx() + field.local_name()->get_string();
+  return (field.node_type() == AST_Decl::NT_sequence) ? (n + "_seq") : n;
+}
+
+std::string FieldInfo::underscored_type_name(UTL_ScopedName* sn, AST_Type& field)
+{
+  std::string n = dds_generator::scoped_helper(sn, "_");
+  if (!field.anonymous()) {
+    return n;
+  }
+  std::size_t i = n.rfind(scope_op);
+  return n.substr(0, i + 2) + at_pfx() + n.substr(i + 2);
+}
+
+std::string FieldInfo::underscore(const std::string& scoped_type)
+{
+  std::string s = scoped_type;
+  for (std::size_t i = s.find(scope_op); i != s.npos; i = s.find(scope_op, i + 2)) {
+    s.replace(i, 2, "_");
+  }
+  return s;
+}
+
+std::string FieldInfo::ref(const std::string& scoped_t, const std::string& underscored_t, const std::string& const_s)
+{
+  return "IDL::DistinctType<" + const_s + scoped_t + ", " + underscored_t + "_tag>";
+}
+
+// for anonymous types
+FieldInfo::FieldInfo(AST_Field& field) :
+  type_(field.field_type()),
+  name_(field.local_name()->get_string()),
+  underscored_(underscored_type_name(type_->name(), *type_))
+{
+  init();
+}
+
+FieldInfo::FieldInfo(UTL_ScopedName* sn, AST_Type* base) :
+  type_(base),
+  underscored_(underscored_type_name(sn, *type_)),
+  scoped_type_(scoped(sn))
+{
+  init();
+}
+
+void FieldInfo::init()
+{
+  act_ = resolveActualType(type_);
+  cls_ = classify(act_);
+  arr_ = AST_Array::narrow_from_decl(type_);
+  seq_ = AST_Sequence::narrow_from_decl(type_);
```

Comment:
```suggestion
  arr_ = dynamic_cast<AST_Array*>(type_);
  seq_ = dynamic_cast<AST_Sequence*>(type_);
```
TAO_IDL narrows are just dynamic casts, so the actual cast should be used when possible. Also you're resolving the type of `type_`, but not using it here?

Reply:
The cast changes were decided as optional in a meeting with Fred and Adam. Both act_, the resolved the type of type_, and its classification cls_ are used in generators.